### DISCUSSION TITLE
Retain relative paths in PSR-4 mapping?

### DISCUSSION THREAD

**Discussion by townxelliot** [composer/composer#11165] [Q&A / Support] [answered]
At present, I have an issue where I am mounting a local directory into docker and get a different vendor/composer/autoload_psr4.php depending on where I run the autoloader from.

The context for this is psalm scanning of our codebase, which makes use of autoload.php. Unfortunately, running the docker containers locally wrecks the autoload.php for the purpose of psalm scanning, due to how we're mounting the local filesystem into the container. I generally have to regenerate autoload.php manually to do a local psalm scan, which makes adding pre-commit hooks difficult (as the state of autoload.php is ambiguous, depending on whether you've run the containers or not).

My directory structure is like this:

```
shared
service-front
  composer.json
```

My service-front/composer.json looks like this:

```
"autoload": {
        "psr-4": {
            ...
            "MakeLogger\\": "../shared/logging/module/MakeLogger/src/",
            ...
        }
    },
```

If I run `composer dumpautoload` from my local machine I get this in autoload_psr4.php:

```
'MakeLogger\\' => array($baseDir . '/../shared/logging/module/MakeLogger/src'),
```

This is what I want. However, if I copy the files from shared and service-front into a docker image with this mapping:

```
service-front => /app
shared => /shared
```

and run composer dumpautoload from inside that container (in the working directory /app), I get this in autoload_psr4.php:

```
'MakeLogger\\' => array('/shared/logging/module/MakeLogger/src')
```

(i.e. the relative path has been collapsed into an absolute path). This is no good, as these paths won't resolve unless I'm running psalm inside the docker container.

I have tried two fixes for this:

1. Modify how files are copied into the image so they aren't hanging off the root `/` directory: `service-front => /app/main` and `shared => /app/shared`, instead of `/app` and `/shared` respectively. If I dump the autoloader now (from /app/main), the relative path is retained (which is what I want). This is quite intrusive and needs a lot of files to be moved around in images, though.
2. Change the Filesystem->normalizePath() method so it doesn't collapse relative paths into absolute ones. The autoloader does what I want if I apply this patch. This is far less intrusive for us, but obviously impacts on composer quite heavily.

Are there other solutions for this? Any possible changes to composer which would be less intrusive but allow this issue to be solved for other developers? Any advice will be gratefully received.

**Comment by townxelliot** [comment]
For reference, this is a similar question: https://github.com/composer/composer/issues/4404. It was closed as an edge case.

**Comment by Seldaek** [comment, marked as answer]
Here is a PR which should fix this only in cases where the dir is top-level `/foo` (i.e. not `/foo/bar`) https://github.com/composer/composer/pull/11169

Can you please try it out see if it solves it for you? The impact seems lower than your patch and hopefully comes without risk of unexpected side effects.

**Reply by townxelliot** [reply]
Excellent, that does the job, thanks. As you say, it retains the `$basedir . '/../shared/'` path instead of collapsing it to `/shared/`, which is ideal for my purposes. I appreciate my patch was quite invasive and I appreciate your rework.
